library/core/src/slice/mod.rs

```diff
@@ -5089,31 +5092,43 @@ impl [f64] {
     }
 }
 
+#[const_trait]
+#[rustc_const_unstable(feature = "const_clone", issue = "142757")]
 trait CloneFromSpec<T> {
-    fn spec_clone_from(&mut self, src: &[T]);
+    fn spec_clone_from(&mut self, src: &[T])
+    where
+        T: [const] Destruct;
 }
 
-impl<T> CloneFromSpec<T> for [T]
+#[rustc_const_unstable(feature = "const_clone", issue = "142757")]
+impl<T> const CloneFromSpec<T> for [T]
 where
-    T: Clone,
+    T: [const] Clone,
 {
     #[track_caller]
-    default fn spec_clone_from(&mut self, src: &[T]) {
+    default fn spec_clone_from(&mut self, src: &[T])
+    where
+        T: [const] Destruct,
+    {
         assert!(self.len() == src.len(), "destination and source slices have different lengths");
         // NOTE: We need to explicitly slice them to the same length
         // to make it easier for the optimizer to elide bounds checking.
         // But since it can't be relied on we also have an explicit specialization for T: Copy.
         let len = self.len();
         let src = &src[..len];
-        for i in 0..len {
-            self[i].clone_from(&src[i]);
+        // FIXME(const_hack): make this a `for idx in 0..self.len()` loop.
+        let mut idx = 0;
+        while idx < self.len() {
```

Comment:
We just had this [discussion about more const hacks](https://rust-lang.zulipchat.com/#narrow/channel/146212-t-compiler.2Fconst-eval/topic/SystemTime.20regression.20.23146228) -- converting for loops to while loops should only happen if there's a really strong motivation. Might be better to just not constify this yet.

Reply:
I understand the sentiment and find it reasonable. But I think that if `for` -> `while`, which is probably the most harmless const hack out there, is expected to be avoided, I think some kind of guideline in style of "no code changes unless there's a really strong motivation" should be created.

Reply:
I guess the question is, is cloning in const useful even if you can't clone slices? I think it may be worth landing the rest first, if only to reduce reviewing overhead here.

In the end this is up to t-libs; @tgross35 what do you think?